### ISSUE TITLE
VST3 folder path

### ISSUE THREAD

**Issue by tank-trax** [attilammagyar/js80p#20]
thanks for making this synth available for Linux systems

for the VST3 a few tweaks are needed

rather than **js80p.vst3** being the name of the binary it should be a folder with subfolders **Contents** and **x86_64-linux**

`js80p.vst3/Contents/x86_64-linux/`

the folder x86_64-linux would have the binary and presets
renaming the binary from js80p.vst3 to js80p.so would be the last step

![image](https://github.com/attilammagyar/js80p/assets/41337033/bae0a885-2310-4a3a-a1b7-473c82f211c6)

**Comment by attilammagyar** [comment]
You're right, the single DLL named `.vst3` is a deprecated format, I'll take a look at the VST3 packaging in the upcoming weeks. (Next week is going to be more busy for me than usual, so it might take some time.) Thanks for letting me know!

**Comment by attilammagyar** [comment]
Well, this is going to be more challenging than I expected: the [VST 3 docs](https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/Locations+Format/Plugin+Format.html#for-the-linux-platform) say that

> MyPlugin.vst3/Contents/XXX-linux   with XXX the architecture name based on the output of command-line "uname -m" (machine hardware) + "-linux"

And indeed, the SDK does seem to use [uname()](https://github.com/steinbergmedia/vst3_public_sdk/blob/e099a33fe9d78295f58f919ec3bb58968dd5752c/source/vst/hosting/module_linux.cpp#L88-L97) on Linux to figure out [where to look for the .so file](https://github.com/steinbergmedia/vst3_public_sdk/blob/e099a33fe9d78295f58f919ec3bb58968dd5752c/source/vst/hosting/module_linux.cpp#L158-L161).

The problem is that when you are running a 32 bit host application on a 64 bit Linux, then `uname()` will not report the machine hardware name that the VST 3 spec expects.

Here's what `strace` shows when `tests/smoke-test/smoke-test.sh` gets to test the 32 bit version of JS80P with the 32 bit version of Reaper on a 64 bit Ubuntu:

    [pid 12634] uname({sysname="Linux", ..., machine= "x86_64", ...}) = 0
    ...
    [pid 12599] openat(AT_FDCWD, "/tmp/js80p-smoke-test/js80p.vst3/Contents/unknown-linux/js80p.so", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
    [pid 12599] stat64("/tmp/js80p-smoke-test/js80p.vst3/Contents/unknown-linux/js80p.so", 0xfff65c00) = -1 ENOENT (No such file or directory)
    [pid 12599] openat(AT_FDCWD, "/tmp/js80p-smoke-test/js80p.vst3", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 18

What's weird is that it doesn't even attempt to find a subdirectory named `x86_64-linux`, instead, it goes straight to `unknown-linux`. Same thing happens if I merge the 64 bit and 32 bit VST 3 into the same bundle: the 32 bit Reaper ignores the `x86_64-linux` option, doesn't even try to load it. I guess it has some expectations about the result of `uname()`, and `x86_64` obviously fails to meet those, so it falls back to `unknown`.

I'm not even sure that it would be a Reaper bug, it seems more like a flaw in the Linux support of the VST 3 SDK, because the plugin architecture to pick for a given host shouldn't depend on the kernel's architecture, but on the architecture of the host application.

I'll need to ask around in Reaper's forums, but now I'm waiting to get my account approved by their mods.

Long story short, until this thing gets sorted out, the deprecated VST 3 format remains, because at least, that one seems to be free from this problem. :disappointed:

**Comment by attilammagyar** [comment]
[v1.8.0](https://github.com/attilammagyar/js80p/releases/tag/v1.8.0) and onwards will come with a merged VST 3 bundle. (Legacy single file VST 3 is still available for older and/or buggy hosts.)

**Comment by tank-trax** [comment]
just thought I'd mention and confirm that the Linux VST3  in the js80p-1_8_0-vst3_bundle.zip works well

cheers

**Comment by attilammagyar** [comment]
Thank you for confirming it, and also thank you for suggesting the idea!
